Layers - State/Queries/ContextMenu/BackwardCompatibility

aldebaran30701 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* Follow the check list items defined by https://github.com/constellation-app/constellation/blob/master/CONTRIBUTING.md#pull-requests
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change
This Pull request wraps the following outstanding PRs into one.
#454 #491 #501 #500 

**State**: #454 
Layers view now keeps state written to graph, and reacts to changes in UI and graph objects to keep pane in sync with graph.

**Queries**: #491 
The queries have test cases and now allow for complete nesting of terms.

**Context Menu**: #501 
The context menu for adding items to the layer now includes a description and clickable sub menu to add elements to.

**Backward Compatibility**: #500 
This includes fixes to attribute fetching to allow older graphs to be used without the need for Layers View specific attributes.

The Layers view now adds elements via draw nodes to the current layer which is being viewed.

<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
See individual PRs
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
It adds functionality, fixes compatibility bugs and allows for a smoother use of the tool.
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an exising file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
It adds functionality, fixes compatibility bugs and allows for a smoother use of the tool.
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
N/A
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
Test cases for queries, step by step UI usage for state, backward compatibility checking done via old V1 graphs. Context menu trialed through usage.
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
#223 
<!-- Link any applicable issues here -->

aldebaran30701 commented:
Change the extension to .star, and this will assist in testing queries.
[LayersViewTest.zip](https://github.com/constellation-app/constellation/files/4626505/LayersViewTest.zip)

A List of steps I used for testing. Formatting messed up with the export
[testing layers.txt](https://github.com/constellation-app/constellation/files/4626510/testing.layers.txt)

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:
```diff
@@ -35,64 +42,156 @@
+        }
+
+        final Graph graph = GraphManager.getDefault().getActiveGraph();
+        if(graph == null){
```
can be merged with above if statement: `if (pane == null || graph == null)`

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:
```diff
@@ -35,64 +42,156 @@
+        }
+
+        final Graph graph = GraphManager.getDefault().getActiveGraph();
+        if(graph == null){
```
can be merged with above if statement: `if (pane == null || graph == null)`

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:
```diff
@@ -35,64 +42,156 @@
+                return;
+            }
+
+            if (pane == null) {
```
can be merged with first if statement: `if (graph == null || pane == null)`

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java:
```diff
@@ -109,47 +115,101 @@ public LayersViewPane(final LayersViewController controller) {
 
-        final TextArea descriptionTextArea = new TextArea();
-        descriptionTextArea.setPrefRowCount(1);
+        final Node descriptionTextArea = new TextArea();
```
Curious as to why this is declared as a Node rather than a TextArea (given all the subsequent typecasting)?

aldebaran30701 replied:
I couldn't manage to get an event to trigger on the `TextArea` when it was inserted in the `GridPane`. So I had to declare it as a Node, then add the event to it.

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewTopComponent.java:
```diff
@@ -76,12 +81,25 @@ protected LayersViewPane createContent() {
+    @Override
+    protected void handleComponentOpened() {
+        layersViewPane.setDefaultLayers();
+        layersViewController.readState();
```
I note that this function as well as the two above it (handleNewGraph, handleGraphOpened) have very similar implementations. Maybe this is something that can be taken advantage of.

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/state/LayersViewConcept.java:
```diff
@@ -0,0 +1,66 @@
+    public static class MetaAttribute {
+
+        private MetaAttribute() {
+            throw new IllegalStateException("Utility class");
```
Change message to "Concept class" or similar

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/query/QueryEvaluator.java:
```diff
@@ -15,110 +15,227 @@
+                currentString = "";
+            } else if (c == '|' && prevChar == '|') {
+                currentString = currentString.substring(0, currentString.lastIndexOf(prevChar));
+                if (!StringUtils.isBlank(currentString)) {
```
There is actually a function `StringUtils.isNotBlank`. I'd use that instead (and do this for the other instances of this in the file)

arcturus2 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/utilities/LayersAddContextMenu.java:
```diff
@@ -0,0 +1,81 @@
+                for (final LayerDescription layer : currentState.getLayers()) {
+                    if (layer.getLayerIndex() > 1) {
+                        final String description = StringUtils.isBlank(layer.getLayerDescription())
+                                ? "[NO DESCRIPTION]" : layer.getLayerDescription();
```
@aldebaran30701 thanks for including this Wishlist item already. Hope you liked the idea too.

A minor comment and you don't have to do it now but things like these "[NO DESCRIPTION]" strings can be made constants because it makes it easier to update and just stands out more like LAYER_MENU etc.

If we want to support multiple languages down the track, it also helps because all of the strings are in one place rather than nested in the code.

aldebaran30701 replied:
I was looking at this thinking the same thing!
No worries, I've got a few suggestions for extra features I can put forward if they seem valuable to users.

aldebaran30701 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/StoreGraph.java:
```diff
@@ -2179,36 +2168,73 @@ private void updateLayerMask(final int attributeId, final GraphElementType eleme
+                transactionFilterBitmaskAttrId != Graph.NOT_FOUND && 
+                transactionLayerVisibilityAttributeId != Graph.NOT_FOUND) {
+            if (attributeId != transactionFilterBitmaskAttrId) {
+                avoidLayerUpdate = true;
```
The idea is that in the case of an attribute other than the Vertex/Transaction's bitmask attribute gets updated - it will not trigger another call to update the layer. `StoreGraph.setIntValue(...)` has a check just before it calls to update layers so that it avoids unnecessary recalculation. Line 2370

aldebaran30701 commented:
> @aldebaran30701 just wanted to acknowledge you've stuck with this change for a long time and accommodated all of the PR feedback in good spirit. It's looking really good.

It's been a journey with plenty of learning. Thanks to all for keeping my code clean and style consistent. A credit to how thorough you all are.

aldebaran30701 commented:
Just added a commit that includes a button to deselect all layers. 
I also added a listener to handle closing of a graph - it now reflects the default case when no graph is shown - prior to this commit it would hang on the last opened graph.
